### PR TITLE
Dynamic template implementation

### PR DESCRIPTION
**Pull request by JackScanlon**
# Issues & Tasks
This PR will resolve several issues, and covers a variety of tasks.

This PR succeeds:
- #1072  

Please see the following issue for further information:
- #1076

### CODE REVIEW

**Comment by ieuans** [comment]
### Frontend
1. [ ] Update API page

### Backend
1. [x] Added validation data to the clinical template, required for API & create/update web interface
2. [ ] API endpoints for all get methods of generic entity
3. [ ] API endpoints for create/update of generic entity

### Misc
1. [x] Added Docker incl. changes to production files
4. [x] Added new README
5. [x] Removed clutter

File: CodeListLibrary_project/clinicalcode/entity_utils/search_utils.py [hunk 1]
```diff
@@ -67,7 +67,10 @@ def get_renderable_entities(request, entity_type=None):
         }
 
     template_ids = list(templates.values_list('id', flat=True))
-    entities = PublishedGenericEntity.objects.filter(template__in=template_ids)
+
+    entities = GenericEntity.objects.filter(template__in=template_ids).all()
+    entities = PublishedGenericEntity.objects.filter(entity__in=entities).order_by('-created').distinct().order_by('id')
```
Comment (JackScanlon) [inline CodeListLibrary_project/clinicalcode/entity_utils/search_utils.py:72]: We need to look into removing PublishedGenericEntity, and all 'published' models in general.

The same could be achieved using the `status` field and `history` objects and it increases the complexity of getting a meaningful read from the DB

File: CodeListLibrary_project/clinicalcode/entity_utils/constants.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        }
+    },
+    'collections': {
+        'title': 'Collections',
+        'active': True,
+        'field_type': 'int_array',
+        'source': 'Tag',
+        'query': 'id',
+        'relative': 'description',
+        'filter': {
```
Comment (JackScanlon) [inline CodeListLibrary_project/clinicalcode/entity_utils/constants.py:187]: We need to look into cleaning up tags and collections, as well as other data models - we use different nomenclature for PK and the name field which increases complexity of querying these fields

File: CodeListLibrary_project/clinicalcode/entity_utils/search_utils.py [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+        entities = search_entities(queryset=entities, search_query=search, order_by_relevance=should_order_search, fuzzy=True, min_threshold=0.1)
+
+    # Reorder by user selection
+    if search_order != constants.ORDER_BY['1']:
+        entities = entities.order_by(search_order.get('clause'))
+    else:
+        if search is None:
+            entities = entities.all().extra(
+                select={'true_id': """CAST(REGEXP_REPLACE(id, '[a-zA-Z]+', '') AS INTEGER)"""}
+            ).order_by('true_id', 'id')
```
Comment (JackScanlon) [inline CodeListLibrary_project/clinicalcode/entity_utils/search_utils.py:325]: We should really change this - needed due to change to ID from int to char field, but not the best solution.

Might be slightly faster to use the following code, but likely still doesn't scale well:
```sql
ORDER BY SUBSTRING(id, '^[0-9]+')::int, SUBSTRING(id, '[^0-9]*$')
```

**Comment by JackScanlon** [comment]
Todo:

Convert '_GenericEntitySearchView_' to ```ListView``` and add caching to improve response time, perf. seems to be related to templating

**Pull request merged**